classes/class-wc-connect-compatibility-wcshipping-packages.php

```diff
@@ -0,0 +1,336 @@
+ *
+ * @since x.x.x
+ */
+class WC_Connect_Compatibility_WCShipping_Packages {
+
+	/**
+	 * Number WCShipping uses to indicate data migration from WCS&T has completed.
+	 *
+	 * @var int
+	 */
+	const WCSHIP_DATA_MIGRATION_COMPLETED = 14;
+
+	/**
+	 * Mapping of WCShipping keys => WCS&T keys.
+	 *
+	 * @var array
+	 */
+	const WCSHIPPING_TO_WCSERVICES_KEY_MAP = array(
+		'boxWeight'  => 'box_weight',
+		'dimensions' => 'inner_dimensions',
+		'maxWeight'  => 'max_weight',
+	);
+
+	/**
+	 * Keys that are allowed in packages mapped to the WCS&T package data format.
+	 *
+	 * Other keys will be removed.
+	 *
+	 * @var array
+	 */
+	const KEYS_USED_BY_WCSERVICES = array(
+		'box_weight',
+		'inner_dimensions',
+		'is_letter',
+		'is_user_defined',
+		'max_weight',
+		'name',
+	);
+
+	/**
+	 * Keys that are allowed in packages mapped to the WCShipping package data format.
+	 *
+	 * Other keys will be removed.
+	 *
+	 * @var array
+	 */
+	const KEYS_USED_BY_WCSHIPPING = array(
+		'boxWeight',
+		'dimensions',
+		'id',
+		'is_user_defined',
+		'maxWeight',
+		'name',
+		'type',
+	);
+
+	/**
+	 * Registers all, some, or no hooks based on store configuration.
+	 *
+	 * @return void
+	 */
+	public static function maybe_enable() {
+		// Don't do anything if WooCommerce Shipping is not active.
+		if ( ! WC_Connect_Loader::is_wc_shipping_activated() ) {
+			return;
+		}
+
+		self::register_rest_controller_hooks();
+
+		$is_migration_to_wcshipping_completed = self::WCSHIP_DATA_MIGRATION_COMPLETED === (int) get_option( 'wcshipping_migration_state' );
+		if ( $is_migration_to_wcshipping_completed ) {
+			self::register_option_overwriting_hooks();
+		}
+	}
+
+	/**
+	 * Enqueue REST controller registration after WCS&T has finished initializing its other controllers.
+	 *
+	 * @return void
+	 */
+	public static function register_rest_controller_hooks() {
+		add_action( 'wcservices_rest_api_init', array( self::class, 'register_wcshipping_compatibility_rest_controller' ) );
+	}
+
+	/**
+	 * Registers hooks intercepting reads/writes to "wc_connect_options".
+	 *
+	 * This is done to replace the keys "packages" and "predefined_packages" with values from WCShipping's options
+	 * after doing some mapping.
+	 *
+	 * @return void
+	 */
+	public static function register_option_overwriting_hooks() {
+		// Intercept reads of "wc_connect_options[packages]" and "wc_connect_options[predefined_packages]".
+		add_filter( 'option_wc_connect_options', array( self::class, 'intercept_packages_read' ) );
+		add_filter( 'option_wc_connect_options', array( self::class, 'intercept_predefined_packages_read' ) );
+
+		// Intercept updates to "wc_connect_options[packages]" and "wc_connect_options[predefined_packages]".
+		add_action( 'update_option_wc_connect_options', array( self::class, 'intercept_packages_update' ), 10, 2 );
+		add_action( 'update_option_wc_connect_options', array( self::class, 'intercept_predefined_packages_update' ), 10, 2 );
```

Comment:
The 2 `update_` functions here will always overwrite WooShipping's packages and predefined packages if the user edits live rate settings under WCS&T's code, regardless if there are any new changes from WCShipping. This may "delete" any new "saved templates" (the blue star) created from WCShipping's side. 

If WCS&T can update `wc_connect_options` and can overwrite `wcshipping_options` but WooShipping can only update `wcshipping_options` but can't modify `wc_connect_options`, then losing saved templates (blue star) is the only problem. We can fix this by adding another level of `array_merge` to combine them. However, if WooShipping can also update `wc_connect_options`, then this will get really messy and we should avoid doing that.

Also, when I clicked "edit" in one of my custom box via http://localhost/wp-admin/admin.php?page=wc-settings&tab=shipping&section=woocommerce-services-settings, these 2 hooks don't get fired. Is that the same for you too?

Reply:
> The 2 update_ functions here will always overwrite WooShipping's packages and predefined packages if the user edits live rate settings under WCS&T's code, regardless if there are any new changes from WCShipping. This may "delete" any new "saved templates" (the blue star) created from WCShipping's side.

They won't! If you look at [this method](https://github.com/Automattic/woocommerce-services/blob/cf4e4556b3d0c1e8799db4a94f96f3c24f1447c5/classes/class-wc-connect-compatibility-wcshipping-packages.php#L158), it first gets WCShip's current options value and then it only replaces the `packages` key.

Whatever WCS&T is writing there is based on current WCShip packages, too, because the read is intercepted as well.

> If WCS&T can update wc_connect_options and can overwrite wcshipping_options but WooShipping can only update wcshipping_options but can't modify wc_connect_options, then losing saved templates (blue star) is the only problem.

WCS&T modifies its entire `wc_connect_options` but the value of the `packages` key is replaced with whatever is in `wcshipping_options[packages]`. In pseudocode, it is doing:

```js
wc_connect_options = {
    ...wc_connect_options,
    packages: wcshipping_options.packages,
};
```

And after a merchant is done modifying the packages in the above object (note they came from the WCShip options), when WCS&T is writing to `wc_connect_options`, it also writes to just the `packages` field of WCShip:

```js
const wcshipping_options = get_from_db( 'wcshipping_options' )

save( 'wc_connect_options', submitted_form_data );

save( 'wcshipping_options', {
    ...wcshipping_options,
    packages: submitted_form_data.packages,
} );
```

> Also, when I clicked "edit" in one of my custom box via http://localhost/wp-admin/admin.php?page=wc-settings&tab=shipping&section=woocommerce-services-settings, these 2 hooks don't get fired. Is that the same for you too?

Please make sure you also have the WC Shipping plugin enabled. Without it, this compatibility layer will not be registered at all.

Reply:
> Whatever WCS&T is writing there is based on current WCShip packages, too, because the read is intercepted as well.
... it also writes to just the packages field of WCShip:

Interesting... that's brilliant! Both read & write are intercepted and but the read-inside-the-write is also intercepted first, so you are reading the WCShipping, turns it to WCS&T settings, then write it, essentially syncing the new database config back to the old ones before writing. 👍 

Ok I need a moment to visualize this...
![image](https://github.com/user-attachments/assets/5422a78f-f4ea-4419-8110-fab50b672ff8)
https://xkcd.com/555/